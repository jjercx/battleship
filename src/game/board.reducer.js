export default (state = null, { type }) => {
  switch (type) {
    default:
      return state;
  }
};
