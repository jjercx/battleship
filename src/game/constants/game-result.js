export const WIN = "WIN";
export const WIN_P1 = "P1 WIN";
export const WIN_P2 = "P2 WIN";
export const TIE = "TIE";
export const LOSE = "LOSE";
