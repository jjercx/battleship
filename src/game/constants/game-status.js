export const IN_PROGRESS = "IN_PROGRESS";
export const LOSE = "LOSE";
export const WIN = "WIN";
export const WIN_P1 = "P1 WIN";
export const WIN_P2 = "P2 WIN";
export const TIE = "TIE";

// TODO: remove IN_PROGRESS and just use ready bool inside gameInfo
// TODO: rename gameStatus into gameResult
