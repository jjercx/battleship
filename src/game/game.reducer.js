import {
  GAME_SETUP,
  GAME_READY,
  SHIP_UPDATE,
} from "app/constants/action-types";

export const getShips = ({ game: { ships } }) => Object.values(ships);

export const getShip = ({ game: { ships } }, shipId) => ships[shipId];

export const getLoading = ({ game: { loading } }) => loading;

export const getCell = ({ game: { board } }, row, col) =>
  board && board[row][col];

const initialState = {
  loading: false,
  board: null,
  ships: {},
};

export default (state = initialState, { type, payload }) => {
  switch (type) {
    case GAME_SETUP:
      return { ...state, loading: true };
    case GAME_READY: {
      const { ships, board } = payload;
      return { ...state, loading: false, board, ships };
    }
    case SHIP_UPDATE: {
      const ship = payload;
      return { ...state, ships: { ...state.ships, ...{ [ship.id]: ship } } };
    }
    default:
      return state;
  }
};
