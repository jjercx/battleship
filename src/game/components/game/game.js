import React from "react";
import styled from "styled-components";
import Stat from "game/components/stat";
import ShipItem from "game/components/ship-item/ship-item";
import { connect } from "react-redux";
import { getShips, getShots, getTurns, getHits } from "game/redux/game.reducer";
import * as actions from "game/redux/game.actions";
import Grid from "game/components/grid";
import { getSize } from "app/redux/game-info.reducer";

const StatPanel = styled.div`
  display: grid;
  grid-template-columns: 1fr 1fr 1fr;
  margin: 10px;
`;

const ShipPanel = styled.div`
  display: grid;
  grid-template-columns: 1fr 1fr;
  grid-template-rows: repeat(5, 1fr);
  margin: 10px;
`;

const Column = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
`;

const Row = styled.div`
  display: flex;
  flex-wrap: wrap-reverse;
  justify-content: center;
`;

export const Game = ({ ships = [], hits, shots, turns, size }) => {
  return (
    <Row>
      <Column>
        <StatPanel>
          <Stat value={hits} title="Hits"></Stat>
          <Stat value={shots} title="Shots"></Stat>
          <Stat value={turns === Infinity ? "âˆž" : turns} title="Turns"></Stat>
        </StatPanel>
        <ShipPanel>
          {ships.map(ship => (
            <ShipItem ship={ship} key={ship.id}></ShipItem>
          ))}
        </ShipPanel>
      </Column>
      <Grid size={size} />
    </Row>
  );
};

const mapStateToProps = state => ({
  ships: getShips(state),
  shots: getShots(state),
  turns: getTurns(state),
  size: getSize(state),
  hits: getHits(state),
});

export default connect(mapStateToProps, actions)(Game);
