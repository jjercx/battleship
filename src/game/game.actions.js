import { GAME_READY, GAME_SETUP } from "app/constants/action-types";

export const gameSetup = ({ size }) => ({
  type: GAME_SETUP,
  payload: size,
});

export const gameReady = ({ ships, board }) => ({
  type: GAME_READY,
  payload: { ships, board },
});
