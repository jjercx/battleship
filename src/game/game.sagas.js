import { takeLatest, put, takeEvery, select } from "redux-saga/effects";
import { GAME_SETUP, TILE_TOUCH } from "app/constants/action-types";
import Game from "game/game";
import { gameReady, shipUpdate } from "game/game.actions";
import { getCell, getShip } from "./game.reducer";

function* gameSetup({ payload: boardSize }) {
  const { ships, board } = Game.setup(boardSize);

  yield put(gameReady({ ships, board }));
}

function* tileTouch({ payload: { row, col } }) {
  const shipId = yield select(getCell, row, col);
  const ship = yield select(getShip, shipId);
  if (ship) {
    ship.hit();
    yield put(shipUpdate({ ship }));
  }
}

export default [
  takeLatest(GAME_SETUP, gameSetup),
  takeEvery(TILE_TOUCH, tileTouch),
];
