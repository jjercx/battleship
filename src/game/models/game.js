import shipData from "game/models/ships.data";
import Ship from "game/models/ship";
import { randomShuffle, randomNumber } from "app/utils/random";
import { EMPTY, DIRECTIONS } from "app/constants/game";

export default class Game {
  static setup(boardSize) {
    const ships = Game.createShipsFromData(shipData.data);
    const board = Game.createEmptyBoard(boardSize);

    Game.fillBoardWithShips(board, ships);

    return { ships, board };
  }

  static createShipsFromData = data => {
    const ships = {};
    data.forEach(({ count, size }) => {
      for (let i = 0; i < count; i++) {
        const ship = new Ship({ size });
        ships[ship.id] = ship;
      }
    });
    return ships;
  };

  static createEmptyBoard = size => {
    const board = [];
    for (let i = 0; i < size; i++) {
      const row = [];
      for (let j = 0; j < size; j++) {
        row.push(EMPTY);
      }
      board.push(row);
    }
    return board;
  };

  static fillBoardWithShips = (board, ships) => {
    Object.values(ships).forEach(ship => {
      const { row, col, h, v } = Game.randomPosition(board, ship);
      for (let k = 0; k < ship.size; k++) {
        board[row + k * h][col + k * v] = ship.id;
      }
    });
  };

  static randomPosition = (board, ship) => {
    while (true) {
      const boardSize = board[0].length;
      const [row, col] = [randomNumber(boardSize), randomNumber(boardSize)];

      if (board[row][col] !== EMPTY) {
        continue;
      }

      const randomDirections = randomShuffle(DIRECTIONS);

      for (let dir of randomDirections) {
        const { h, v } = dir;
        let valid = true;

        for (let k = 0; k < ship.size; k++) {
          const newRow = row + k * h;
          const newCol = col + k * v;

          if (newRow >= boardSize || newRow < 0) {
            valid = false;
            break;
          }

          if (newCol >= boardSize || newCol < 0) {
            valid = false;
            break;
          }

          if (board[newRow][newCol] !== EMPTY) {
            valid = false;
            break;
          }
        }
        if (valid) {
          return { row, col, h, v };
        }
      }
    }
  };
}
