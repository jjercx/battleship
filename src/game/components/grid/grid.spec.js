import React from "react";
import Grid from "./grid";
import renderer from "react-test-renderer";
import { EMPTY } from "app/constants/game";

describe("<Grid />", () => {
  describe("with size", () => {
    it("matches snapshot", () => {
      const tree = renderer.create(<Grid />).toJSON();

      expect(tree).toMatchSnapshot();
    });
  });

  describe("without size", () => {
    it("matches snapshot", () => {
      const store = {
        game: {
          board: [[EMPTY]],
        },
      };
      const tree = renderer.create(<Grid store={store} />).toJSON();

      expect(tree).toMatchSnapshot();
    });
  });
});
