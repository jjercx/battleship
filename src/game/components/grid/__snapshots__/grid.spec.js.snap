// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Grid /> with size matches snapshot 1`] = `null`;

exports[`<Grid /> without size matches snapshot 1`] = `null`;
